File: src/main/dart/g0701_0800/s0739_daily_temperatures/Solution.dart
// #Medium #Top_100_Liked_Questions #Array #Stack #Monotonic_Stack #LeetCode_75_Monotonic_Stack
// #Programming_Skills_II_Day_6 #Big_O_Time_O(n)_Space_O(n)
// #2024_10_12_Time_429_ms_(80.43%)_Space_189.2_MB_(54.35%)

class Solution {
  List<int> dailyTemperatures(List<int> temperatures) {
    List<int> sol = List.filled(temperatures.length, 0);
    sol[temperatures.length - 1] = 0;

    for (int i = sol.length - 2; i >= 0; i--) {
      int j = i + 1;
      while (j < sol.length) { // Use '<' instead of '<='
        if (temperatures[i] < temperatures[j]) {
          sol[i] = j - i;
          break;
        } else {
          if (sol[j] == 0) {
            break;
          }
          j = j + sol[j];
        }
      }
    }
    return sol;
  }
}

File: src/main/python/g0701_0800/s0739_daily_temperatures/Solution.py
# #Medium #Top_100_Liked_Questions #Array #Stack #Monotonic_Stack #LeetCode_75_Monotonic_Stack
# #Programming_Skills_II_Day_6 #Big_O_Time_O(n)_Space_O(n)
# #2024_06_07_Time_864_ms_(87.84%)_Space_30.8_MB_(64.29%)

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        sol = [0] * len(temperatures)
        for i in range(len(temperatures) - 2, -1, -1):
            j = i + 1
            while j < len(temperatures):
                if temperatures[i] < temperatures[j]:
                    sol[i] = j - i
                    break
                elif sol[j] == 0:
                    break
                else:
                    j += sol[j]
        return sol
